client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -46,6 +52,9 @@
     private final StatementClient client;
     private final String queryId;
 
+    @GuardedBy("this")
+    private final CopyOnWriteArrayList<Runnable> afterClose = new CopyOnWriteArrayList<>();
```

Comment:
We shouldn't need `CopyOnWriteArrayList` since the access is always synchronized.